test/e2e/clusterctl_upgrade.go

```diff
@@ -65,10 +65,12 @@ type ClusterctlUpgradeSpecInput struct {
 	// provider contract to use to initialise the secondary management cluster, e.g. `v1alpha3`
 	InitWithProvidersContract string
 	SkipCleanup               bool
+	PreCreateCluster          func(managementClusterProxy framework.ClusterProxy, managementClusterNamespace string)
```

Comment:
It would be great if this could align with the functions added in https://github.com/kubernetes-sigs/cluster-api/pull/6636/files e.g. https://github.com/kubernetes-sigs/cluster-api/blob/e5f7e844eb28df8987e128b944eb227099cf4344/test/e2e/cluster_upgrade_runtimesdk.go#L182 

That would mean making this an anonymous function that can then be passed in with any args depending on the needs of the providers.

Reply:
I have used `PreWaitForCluster` to call hook when creating the management cluster but when creating the test workload we are just generating a template and applying it I just kept the simple call for the hook in an if condition

Reply:
+1 for making it an anonymous function. Say I need to know the name of the cluster also, not just the namespace. It would be impossible with the current spec, but if we change it to an anonymous function I can add that as needed.

Reply:
Emm, making `PreCreateCluster` anonymous without explicit paramters does not seem helping in my case because the main thing behind setting this hook is to get the random namespace generated before creating the cluster. And I do not see a way to to get the name space in this hook without setting it explicitly! Do you have any suggestions @killianmuldoon

Reply:
Good point. I didn't think about that...

Reply:
> That would mean making this an anonymous function that can then be passed in with any args depending on the needs of the providers.

I'm slightly confused how that can work. If we define the `PreCreateCluster` as a `func()` without parameters folks can only set funcs without parameters. How could they pass in "with any args depending on the needs"?

Maybe I'm just misunderstanding the suggestion